File: python/repl/monkey/builtins.py
from object import Builtin
from object import Integer
from object import String
from object import ArrayObject

def len_builtin(*args, **kwargs):
    if len(args) != 1:
        return None

    if type(args[0]) == String:
        return Integer(Value=len(args[0].value))
    if type(args[0]) == ArrayObject:
        return Integer(Value=len(args[0].elements))
    return None

builtins_map = {
    'len': Builtin(builtin_function=len_builtin)
}

File: cpp/vm/advm/builtins.cpp
#include "builtins.h"

Builtin::Builtin(std::string n, BuiltinFunction bf) {
    name = n;
    //builtin_function = bf;
    builtin_object = new AdObjectBuiltin(bf);
}

AdObject* len_builtin(std::vector<AdObject*> args, GarbageCollector* gc) {
    AdObject *obj = args.at(0);
    if (obj->type == OT_LIST) {
        AdObjectInteger *result = new AdObjectInteger(((AdObjectList *)obj)->elements.size());
        gc->addObject(result);
        return result;
    }
    return NULL;
}

AdObject* print_builtin(std::vector<AdObject*> args, GarbageCollector* gc) {
    AdObject *obj = args.at(0);
    std::cout << obj->inspect();
    return NULL;
}

AdObject* println_builtin(std::vector<AdObject*> args, GarbageCollector* gc) {
    AdObject *obj = args.at(0);
    std::cout << obj->inspect() << '\n';
    return NULL;
}

void finalize_free_all_builtin_objects() {
    for (auto &builtin : builtins) {
        free_memory_AdObject(builtin.builtin_object);
    }
}